### ISSUE TITLE
context.report() does not handle 'my message {{0}}'

### ISSUE THREAD

**Issue by Skalman** [eslint/eslint#2746]
I'd like the following to work, when writing a custom rule:
```js
context.report(node, 'my message {{0}}', ['testing'])
// should result in "my message testing"
```

The reason is this line:
https://github.com/eslint/eslint/blob/7a3c545272821a2aa47bfd87b0c01ba48812616f/lib/eslint.js#L779
```js
var rx = new RegExp("{{" + escapeRegExp(key) + "}}", "g");

// should be replaced by
var rx = new RegExp(escapeRegExp("{{" + key + "}}"), "g");
```

**Comment by gyandeeps** [comment]
why don'y you do it like this:

```js
context.report(node, 'my message {{name}}', {name: 'testing'})
```

**Comment by nzakas** [comment]
I'm sorry, this is working as intended and won't be making a change.

**Comment by Skalman** [comment]
@gyandeeps Because I'm automatically generating the key(s), see below.
@nzakas Are you sure? Is there a reason not to support numeric keys? And why would you want any numeric keys to match weird things?
```js
var match = /console\.(log|debug)/.exec(str);
if (match)
   context.report(node, 'Your code contains {{0}}', match);

// Generated regexp:
/{{0}}/ == /[\{]{0}\}/ == /\}/

context.report(node, 'my message {{name}}', {0: '!', name: 'testing'});
// generates 'my message {{name!!'
```

I really don't see a reason to keep this bug.

**Comment by michaelficarra** [comment]
:+1: This is a bug and we should fix it.

**Comment by nzakas** [comment]
This is working as intended. The last argument cannot be an array.

**Comment by nzakas** [comment]
Sorry, was really thrown off by your using an array as the last argument.  Please be more descriptive when filing issues, it really saves me a lot of time as I triage a ton every day. 

Let's use this as the test:

```
context.report(node, 'my message {{name}}{{0}}', {0: '!', name: 'testing'});
```

**Comment by Skalman** [comment]
@gyandeeps Thanks for implementing the fix!
@nzakas I'm sorry for the confusion. I thought I made myself clear, but I'll try to be more explicit in the future.
